src/platform/intel/cavs/lib/pm_memory.c

```diff
@@ -0,0 +1,202 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2019 Intel Corporation. All rights reserved.
+ *
+ * Author: Jakub Dabek <jakub.dabek@intel.com>
+ */
+
+#include <sof/lib/pm_memory.h>
+#include <sof/trace/trace.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/shim.h>
+#include <sof/lib/wait.h>
+#include <user/trace.h>
+#include <stdint.h>
+
+#define trace_mem_pm(__e, ...) \
+	trace_event(TRACE_CLASS_MEM, __e, ##__VA_ARGS__)
+#if CAVS_VERSION >= CAVS_VERSION_1_8
+
+#define EBB_SEGMENT_SIZE_ZERO_BASE (EBB_SEGMENT_SIZE - 1)
+/**
+ * \brief Read masks from hw registers and prepare new masks from data
+ * \param[in] ebb_data All masks data required for calculations.
+ */
+static void set_bank_masks(struct ebb_data *operation_data)
+{
+	/* Bit is set for gated banks in  HSPGISTS register so
+	 * we negate it to have consistent 1 = powered 0 = gated
+	 * convention since HSPGISTS registers have it other way
+	 */
+	operation_data->ebb_current_mask0 = ~io_reg_read(HSPGISTS0);
+	operation_data->ebb_current_mask1 = ~io_reg_read(HSPGISTS1);
+
+	/* Calculate mask of bits (banks) to set to powered */
+	operation_data->change_mask0 = SET_BITS
+		(operation_data->ending_bank_id,
+		 operation_data->start_bank_id,
+		 MASK(operation_data->ending_bank_id, 0));
+	operation_data->change_mask1 = SET_BITS
+		(operation_data->ending_bank_id_high,
+		 operation_data->start_bank_id_high,
+		 MASK(operation_data->ending_bank_id_high, 0));
+}
+
+/**
+ * \brief Calculates new mask to write to hw register
+ * \param[in] ebb_data All masks data required for calculations.
+ * \param[in] bit Boolean deciding banks desired state (1 powered 0 gated).
+ */
+static void calculate_new_masks(struct ebb_data *operation_data, uint32_t bit)
+{
+	/* check if either start or end bank is in different segment
+	 * and treat it on case by case basis first only banks
+	 * the segments above EBB_SEGMENT_SIZE are managed
+	 * (separate hw register has to be used for ebb
+	 * numbered over 32 different ebb segment)
+	 */
+	if (operation_data->start_bank_id > EBB_SEGMENT_SIZE_ZERO_BASE &&
+	    operation_data->ending_bank_id > EBB_SEGMENT_SIZE_ZERO_BASE) {
+		/* Set the bits for all powered banks operations
+		 * depend whether we turn on or off the memory
+		 */
+		operation_data->ebb_new_mask0 =
+			operation_data->ebb_current_mask0;
+		operation_data->ebb_new_mask1 = bit ?
+			operation_data->ebb_current_mask1
+			| operation_data->change_mask1 :
+			operation_data->ebb_current_mask1
+			& ~operation_data->change_mask1;
+
+	/* second check if managing spans above and bellow EBB_SEGMENT_SIZE
```

Comment:
Typo: below.

Reply:
bellow
verb
1.
(of a person or animal) emit a deep loud roar, typically in pain or anger.
"he bellowed in agony"

;) Fixed